### ISSUE TITLE
Improve 'files in use by other process' message

### ISSUE THREAD

**Issue by coffeemug** [rethinkdb/rethinkdb#85]
If the user starts up a rethinkdb process and points it to a data directory that's already in use by another rethinkdb process, we give them an error message and quit. It turns out that users don't always trust us, so we should do a little better -- give the pid of the process that's actually using the file. This way they'll know exactly what's going on without having to question the message or hunting around for the rogue process.

**Comment by myth17** [comment]
Hello

I am a newbie and would love to contribute to rethinkdb.

For solving this problem : Can we simply write the process PID whenever acquiring lock into any file in the rethinkdb_data directory and compare it with the PID of the current process at startup?

Looking at the discussion (based on a solution using process creation time and not simple pid comparison), here is my understanding on how we can solve the problem:

	1. Write and Update process PID whenever acquiring lock into any file in the rethinkdb_data. 
	2. Check this last process creation time and compare it with the current process creation time.

To implement Step 1 (Getting Process Id) in Linux and OSX :
	We can use getpid() from <unistd.h> as it seems to be available in both OS.

To implement Step 2 (Finding process creation time) in Linux :
	Read proc file system and calculate the time : There seem to be some good existing solution using: /proc/uptime and /proc/PID/stat

To implement Step 2 in OSX :
	I could not find any solution for this, seems like I could try looking into the source of top or ps for help.
	
Would love to hear suggestions from experienced devs. :)

**Comment by danielmewes** [comment]
Hi @myth17, I'd suggest a slightly simpler implementation. First of all, you might want to take a look at `directory_lock_t::directory_lock_t()` in this file https://github.com/rethinkdb/rethinkdb/blob/next/src/clustering/administration/main/directory_lock.cc#L46 .
The relevant line for the type of error that we're talking about here is this:
```
if (flock(directory_fd.get(), LOCK_EX | LOCK_NB) != 0) {
    throw directory_locked_exc_t(directory_path);
}
```

You can see that we're using the `flock` library function. I think this is a good idea and we should keep that. Among others, it handles the case where you kill the process automatically and releases the lock with the termination of the process.

So the only thing we really need to do is improve the message to mention the PID of the process that is holding the lock (the message is currently defined in directory_locked_exc_t in directory_lock.hpp). The following is a simple solution to this, and I think it's good enough:
1. As soon as a RethinkDB instance has acquired the lock, it gets its own PID (like you describe) and writes it into a special file in the directory, for example `last_locked_by.pid` or something.
2. If the `flock` fails (in a second RethinkDB process), it reads that file and incorporates that PID into the error message, for example it could say:
"Directory '%s' is already in use by process %d. Is another instance of rethinkdb using it?"

Does that sound reasonable to you?

You could check `/proc/PID/stat` in addition to this to make sure that the process is actually still running, and possibly retrieve some additional information about it. I think that's strictly optional though and probably not worth the hassle.

**Comment by myth17** [comment]
I observed that there is some existing code dealing with PID files in `command_line.cc` : https://github.com/rethinkdb/rethinkdb/blob/next/src/clustering/administration/main/command_line.cc

Is it okay if similar code gets duplicated in `directory_lock.cc` ?

**Comment by danielmewes** [comment]
@myth17 good catch. Could you try to refactor the pid file functions from `command_line.cc` and put them in their own `.h` + `.cc` file instead?

**Comment by thelinuxlich** [comment]
@srh I think this can be closed, due to #4745
